File: sdk/src/profiles/graph/apply.js
// weavepack-graph — delta application (applyChain).
//
// Operates on decoded graph state. No byte-level work here.
//
// Graph state:
//   {
//     schemaHash: Uint8Array,
//     nodes: Map<string, { label: string|null, props: Map<number, { ctype, value }> }>,
//     edges: Map<string, { label: string|null, src: BigInt, dst: BigInt,
//                          props: Map<number, { ctype, value }> }>,
//   }
//
// Map keys are nid.toString() / eid.toString() for BigInt identity.
//
// Profile isolation: imports only from ./types.js.

import { OP, PATH_KIND, CTYPE } from "./types.js"

// ── Helpers ────────────────────────────────────────────────────────────────

function cloneState(state) {
  const nodes = new Map()
  for (const [k, v] of state.nodes) {
    nodes.set(k, { label: v.label, props: new Map(v.props) })
  }
  const edges = new Map()
  for (const [k, v] of state.edges) {
    edges.set(k, { label: v.label, src: v.src, dst: v.dst, props: new Map(v.props) })
  }
  return { schemaHash: state.schemaHash.slice(), nodes, edges }
}

// Extract all nids/eids from a block into the state.
function populateNodesFromBlock(nodes, block) {
  const { label, nids, columns } = block
  for (let i = 0; i < nids.length; i++) {
    const key = String(nids[i])
    if (nodes.has(key))
      throw new Error(`duplicate_element_id: nid ${nids[i]} already exists`)
    const props = new Map()
    for (const col of columns) {
      if (col.values[i] !== null && col.values[i] !== undefined) {
        props.set(col.colId, { ctype: col.ctype, value: col.values[i] })
      }
    }
    nodes.set(key, { label: label ?? null, props })
  }
}

function populateEdgesFromBlock(edges, block) {
  const { label, eids, srcs, dsts, columns } = block
  for (let i = 0; i < eids.length; i++) {
    const key = String(eids[i])
    if (edges.has(key))
      throw new Error(`duplicate_element_id: eid ${eids[i]} already exists`)
    const props = new Map()
    for (const col of columns) {
      if (col.values[i] !== null && col.values[i] !== undefined) {
        props.set(col.colId, { ctype: col.ctype, value: col.values[i] })
      }
    }
    edges.set(key, { label: label ?? null, src: srcs[i], dst: dsts[i], props })
  }
}

// ── Path resolution for prop_set ───────────────────────────────────────────

function resolveElementColId(state, path) {
  switch (path.kind) {
    case PATH_KIND.NODE_COL:
      return { elementKey: String(path.nid), isNode: true, colId: path.colId }
    case PATH_KIND.EDGE_COL:
      return { elementKey: String(path.eid), isNode: false, colId: path.colId }
    case PATH_KIND.NODE:
      // node/N without col_id: in schemaless prop_set, use col_id from op context
      return { elementKey: String(path.nid), isNode: true, colId: null }
    case PATH_KIND.EDGE:
      return { elementKey: String(path.eid), isNode: false, colId: null }
    case PATH_KIND.NODE_PROP:
      // prop_name requires schema; not resolved here — treated as named prop
      return { elementKey: String(path.nid), isNode: true, colId: null, prop: path.prop }
    case PATH_KIND.EDGE_PROP:
      return { elementKey: String(path.eid), isNode: false, colId: null, prop: path.prop }
    default:
      throw new Error(`prop_set: path kind ${path.kind} is not a valid element address for prop_set`)
  }
}

// ── Op application ─────────────────────────────────────────────────────────

function applyOp(state, op) {
  state = cloneState(state)

  switch (op.op) {
    case OP.NODE_INSERT: {
      populateNodesFromBlock(state.nodes, op.block)
      break
    }

    case OP.NODE_DELETE: {
      for (const nid of op.nids) {
        const key = String(BigInt(nid))
        state.nodes.delete(key)
        // Implicitly remove incident edges.
        const nidBig = BigInt(nid)
        for (const [ek, ev] of state.edges) {
          if (ev.src === nidBig || ev.dst === nidBig) state.edges.delete(ek)
        }
      }
      break
    }

    case OP.EDGE_INSERT: {
      populateEdgesFromBlock(state.edges, op.block)
      break
    }

    case OP.EDGE_DELETE: {
      for (const eid of op.eids) state.edges.delete(String(BigInt(eid)))
      break
    }

    case OP.PROP_SET: {
      const { elementKey, isNode, colId, prop } = resolveElementColId(state, op.path)
      const target = isNode ? state.nodes.get(elementKey) : state.edges.get(elementKey)
      if (!target)
        throw new Error(`element_not_found: ${isNode ? 'node' : 'edge'} ${elementKey} not found`)
      if (op.isNull) {
        // Set to null: remove the property entry.
        if (colId !== null) target.props.delete(colId)
      } else {
        if (colId !== null) {
          target.props.set(colId, { ctype: op.ctype, value: op.value })
        } else if (prop !== undefined) {
          // Named prop in schemaless mode: use prop name as a pseudo-colId string.
          target.props.set(prop, { ctype: op.ctype, value: op.value })
        }
      }
      break
    }

    case OP.SUBGRAPH_REPLACE: {
      const label = op.label ?? null
      // Step 1: remove all nodes with matching label.
      for (const [k, v] of state.nodes) {
        if (v.label === label) state.nodes.delete(k)
      }
      // Step 2: remove all edges with matching label (+ cascade from step 1).
      for (const [k, v] of state.edges) {
        if (v.label === label) state.edges.delete(k)
      }
      // Step 3: insert replacement node block.
      if (op.nodeBlock) populateNodesFromBlock(state.nodes, op.nodeBlock)
      // Step 4: insert replacement edge block.
      if (op.edgeBlock) populateEdgesFromBlock(state.edges, op.edgeBlock)
      break
    }

    default:
      throw new Error(`unknown_delta_op: op code ${op.op}`)
  }

  return state
}

// ── Public ─────────────────────────────────────────────────────────────────

// Build initial state from a decoded graph document.
export function initState(graph) {
  const state = {
    schemaHash: graph.schemaHash.slice(),
    nodes: new Map(),
    edges: new Map(),
  }
  for (const blk of (graph.blocks ?? [])) {
    if (blk.type === 'node') {
      populateNodesFromBlock(state.nodes, blk)
    } else if (blk.type === 'edge') {
      populateEdgesFromBlock(state.edges, blk)
    }
  }
  return state
}

export function applyChain(state, ops) {
  let s = state
  for (const op of ops) s = applyOp(s, op)
  return s
}

File: impl/python/weavepack_graph/apply.py
"""weavepack-graph — delta application (init_state, apply_chain).

Graph state:
  {
    'nodes': dict[str, {'label': str|None, 'props': dict[int|str, {'ctype': int, 'value': any}]}],
    'edges': dict[str, {'label': str|None, 'src': int, 'dst': int,
                         'props': dict[int|str, {'ctype': int, 'value': any}]}],
  }

Map keys are str(nid) / str(eid) for consistent identity.

Profile isolation: imports only from .types.
"""

import copy
from .types import OP, PATH_KIND


def _populate_nodes_from_block(nodes: dict, block: dict):
    label = block.get("label")
    nids = block.get("nids") or []
    columns = block.get("columns") or []
    for i, nid in enumerate(nids):
        key = str(nid)
        if key in nodes:
            raise ValueError(f"duplicate_element_id: nid {nid} already exists")
        props = {}
        for col in columns:
            v = col["values"][i]
            if v is not None:
                props[col["colId"]] = {"ctype": col["ctype"], "value": v}
        nodes[key] = {"label": label, "props": props}


def _populate_edges_from_block(edges: dict, block: dict):
    label = block.get("label")
    eids = block.get("eids") or []
    srcs = block.get("srcs") or []
    dsts = block.get("dsts") or []
    columns = block.get("columns") or []
    for i, eid in enumerate(eids):
        key = str(eid)
        if key in edges:
            raise ValueError(f"duplicate_element_id: eid {eid} already exists")
        props = {}
        for col in columns:
            v = col["values"][i]
            if v is not None:
                props[col["colId"]] = {"ctype": col["ctype"], "value": v}
        edges[key] = {"label": label, "src": srcs[i], "dst": dsts[i], "props": props}


def init_state(graph: dict) -> dict:
    """Build initial state from a decoded graph document."""
    state = {"nodes": {}, "edges": {}}
    for blk in (graph.get("blocks") or []):
        if blk.get("type") == "node":
            _populate_nodes_from_block(state["nodes"], blk)
        elif blk.get("type") == "edge":
            _populate_edges_from_block(state["edges"], blk)
    return state


def _resolve_element(state: dict, path: dict):
    """Return (element_dict, key, is_node, col_key) for a prop_set path."""
    kind = path["kind"]
    if kind == PATH_KIND.NODE_COL:
        key = str(path["nid"])
        return state["nodes"].get(key), key, True, path["colId"]
    elif kind == PATH_KIND.EDGE_COL:
        key = str(path["eid"])
        return state["edges"].get(key), key, False, path["colId"]
    elif kind == PATH_KIND.NODE:
        key = str(path["nid"])
        return state["nodes"].get(key), key, True, None
    elif kind == PATH_KIND.EDGE:
        key = str(path["eid"])
        return state["edges"].get(key), key, False, None
    elif kind == PATH_KIND.NODE_PROP:
        key = str(path["nid"])
        return state["nodes"].get(key), key, True, path.get("prop")
    elif kind == PATH_KIND.EDGE_PROP:
        key = str(path["eid"])
        return state["edges"].get(key), key, False, path.get("prop")
    else:
        raise ValueError(
            f"prop_set: path kind {kind} is not a valid element address for prop_set"
        )


def apply_op(state: dict, op: dict) -> dict:
    """Apply a single op to a graph state. Returns a new state (deep-copied)."""
    state = copy.deepcopy(state)
    op_code = op["op"]

    if op_code == OP.NODE_INSERT:
        _populate_nodes_from_block(state["nodes"], op["block"])

    elif op_code == OP.NODE_DELETE:
        for nid in (op.get("nids") or []):
            key = str(int(nid))
            state["nodes"].pop(key, None)
            nid_int = int(nid)
            for ek in list(state["edges"]):
                ev = state["edges"][ek]
                if ev["src"] == nid_int or ev["dst"] == nid_int:
                    del state["edges"][ek]

    elif op_code == OP.EDGE_INSERT:
        _populate_edges_from_block(state["edges"], op["block"])

    elif op_code == OP.EDGE_DELETE:
        for eid in (op.get("eids") or []):
            state["edges"].pop(str(int(eid)), None)

    elif op_code == OP.PROP_SET:
        target, key, is_node, col_key = _resolve_element(state, op["path"])
        if target is None:
            elem_type = "node" if is_node else "edge"
            raise ValueError(f"element_not_found: {elem_type} {key} not found")
        is_null = op.get("isNull", False)
        if is_null:
            if col_key is not None:
                target["props"].pop(col_key, None)
        else:
            if col_key is not None:
                target["props"][col_key] = {"ctype": op["ctype"], "value": op.get("value")}

    elif op_code == OP.SUBGRAPH_REPLACE:
        label = op.get("label")
        for k in list(state["nodes"]):
            if state["nodes"][k]["label"] == label:
                del state["nodes"][k]
        for k in list(state["edges"]):
            if state["edges"][k]["label"] == label:
                del state["edges"][k]
        if op.get("nodeBlock") is not None:
            _populate_nodes_from_block(state["nodes"], op["nodeBlock"])
        if op.get("edgeBlock") is not None:
            _populate_edges_from_block(state["edges"], op["edgeBlock"])

    else:
        raise ValueError(f"unknown_delta_op: op code {op_code}")

    return state


def apply_chain(state: dict, ops: list) -> dict:
    """Apply a list of ops in order. Returns the final graph state."""
    for op in ops:
        state = apply_op(state, op)
    return state
